File: Ghidra/Features/BSim/ghidra_scripts/QueryWithFiltersScript.py
## ###
#  IP: GHIDRA
# 
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  
#       http://www.apache.org/licenses/LICENSE-2.0
#  
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
##
# Advanced example of BSim querying
# @category BSim.python

import ghidra.query.facade.SimilarFunctionQueryService as SimilarFunctionQueryService
import ghidra.query.facade.SFQueryInfo as SFQueryInfo
import ghidra.query.FunctionDatabase as FunctionDatabase
import ghidra.query.facade.QueryDatabaseException as QueryDatabaseException
import java.util.HashSet as HashSet
import ghidra.app.plugin.core.query.QueryNearestRow as QueryNearestRow
import java.util.function.BiPredicate as BiPredicate
import ghidra.query.protocol.FilterTemplate as FilterTemplate
import ghidra.app.plugin.core.query.ExecutableResult as ExecutableResult
import java.util.Comparator as Comparator
import java.util.Arrays as Arrays
import java.lang.Double as Double

#Query thresholds
MAX_NUM_FUNCTIONS = 100
SIMILARITY_BOUND = 0.7
SIGNIFICANCE_BOUND = 0.0

#limit the number of results displayed
NUM_EXES_TO_DISPLAY = 10

#for prefiltering: this number will be used to filter out small functions
SELF_SIGNIFICANCE_BOUND = 40.0

def run():

    #get the set of functions to query
    funcsToQuery = getFunctionsToQuery()

    #sets up the object required for querying the database
    queryService = SimilarFunctionQueryService(currentProgram)    
    queryInfo = SFQueryInfo(funcsToQuery)
    bsimFilter = queryInfo.getBsimFilter()

    #sets the query parameters.  
    #change the defined constants to control how fuzzy of
    #a match you're willing to accept, and the maximum number
    #of matches to return for each function
    queryInfo.setMaximumResults(MAX_NUM_FUNCTIONS)
    queryInfo.setSimilarityThreshold(SIMILARITY_BOUND)
    queryInfo.setSignificanceThreshold(SIGNIFICANCE_BOUND)

    #add the prefilters
    setPrefilters(queryService, queryInfo)

    #add a filter on the date
    addBsimFilter(bsimFilter, FilterTemplate.DateLater(""), "01/01/1776")

    #add a filter with multiple values.  Since this is an "Equal" filter, the results are OR'd together
    #so a given executable will pass the main filter if it passes at least one of the subfilters
    addBsimFilter(bsimFilter, FilterTemplate.ArchEquals(),"x86:LE:64:default, x86:LE:32:default, ARM:LE:32:v4")   

    #now add a "notequal" filter
    #to pass, the compiler can't be windows and it can't be foo_compiler
    addBsimFilter(bsimFilter,FilterTemplate.CompNotEqual(),"windows, foo_compiler")
 
    #establish a connection to the BSim database
    try:
        dbUrl = askString("","Enter the URL of the BSim database:", "ghidra://localhost/bsimDB")
        queryService.initializeDatabase(dbUrl)
        error = queryService.getDatabase().getLastError()
        if error is not None and (error.category is ErrorCategory.Nodatabase):
            print "Database [%s] cannot be found (does it exist?)" % dbUrl
            return
    except QueryDatabaseException as e:
        print e.getMessage()
        return

    resultRows = executeQuery(queryService,queryInfo)
    printFunctionQueryResults(resultRows, "\nFunction-level results before filtering")

    #now add some post-query filters, which filters the result set returned by the previous query

    addBsimFilter(bsimFilter, FilterTemplate.Md5NotEqual(), currentProgram.getExecutableMD5())
    addBsimFilter(bsimFilter, FilterTemplate.CompilerEquals(), "gcc")
    addBsimFilter(bsimFilter, FilterTemplate.FunctionTagTemplate("KNOWN_LIBRARY", queryService), "false")

    #apply the filters and print the results
    filteredRows = QueryNearestRow.filterMatchRows(bsimFilter, resultRows)
    printFunctionQueryResults(filteredRows, "\nFunction-level results after filtering")
    printExecutableInformation(filteredRows)
    return


#collect the functions to query from currentProgram
def getFunctionsToQuery():
    functions = HashSet();
    fIter = currentProgram.getFunctionManager().getFunctionsNoStubs(True)
    for func in fIter:
        functions.add(func.getSymbol())
    return functions
    
#query the database
def executeQuery(queryService,queryInfo):
    queryResults = queryService.querySimilarFunctions(queryInfo,monitor)
    resultRows = QueryNearestRow.generate(queryResults.getSimilarityResults(),currentProgram)
    return resultRows

def printFunctionQueryResults(resultRows, title):
    print "%s: %d\n\n" % (title, resultRows.size())
    for row in resultRows:
        print "  queried function: %s" % row.getOriginalFunctionDescription().getFunctionName()
        print "  matching function: %s" % row.getMatchFunctionDescription().getFunctionName()
        print "  executable of matching function: %s" % row.getMatchFunctionDescription().getExecutableRecord().getNameExec()
        print "  similarity: %f" % row.getSimilarity()
        print "  significance: %f\n" % row.getSignificance()

#Prefilters are used to filter out functions before sending a query to the database
#A typical use case would be to collect all functions in a binary, then use a 
#prefilter to remove the functions with low self-significance (which is the 
#"BSim way" to remove small functions)
def setPrefilters(queryService, queryInfo):
    preFilter = queryInfo.getPreFilter();
    selfSigFilter = ExampleFilter(queryService)
    preFilter.addPredicate(selfSigFilter)

class ExampleFilter(BiPredicate):
   
    def __init__(self, queryService):
        self.queryService = queryService

    def test(self,program, fdesc):
        return self.queryService.getLSHVectorFactory().getSelfSignificance(fdesc.getSignatureRecord().getLSHVector()) >= SELF_SIGNIFICANCE_BOUND

def addBsimFilter(bsimFilter, filterTemplate, values):
    for value in values.split(","):
        if len(value.strip()) > 0:
            bsimFilter.addAtom(filterTemplate, value.strip(), FilterTemplate.Blank())

#calls the methods to aggregate executable-level information about the matches 
def printExecutableInformation(filteredRows):
    execrows = ExecutableResult.generateFromMatchRows(filteredRows)
    results = execrows.toArray()
    sorter = Sorter()
    Arrays.sort(results,sorter) 
    print "Executable-level results:"
    numExes = min(len(results),NUM_EXES_TO_DISPLAY)
    for i in range (numExes):
        print "  MD5: %s" % results[i].getExecutableRecord().getMd5()
        print "  Executable Name: %s" % results[i].getExecutableRecord().getNameExec()
        print "  Function Count: %d" % results[i].getFunctionCount()
        print "  Significance Sum: %f\n" % results[i].getSignificanceSum()
    return

class Sorter(Comparator):

    def __init__(self):
        return
    
    def compare(self,o1,o2):
        return Double.compare(o2.getSignificanceSum(), o1.getSignificanceSum()) 



run()

File: Ghidra/Features/BSim/ghidra_scripts/QueryWithFiltersScript.java
/* ###
 * IP: GHIDRA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
//Example of a script to perform a more involved BSim query.
//@category BSim
import java.util.*;
import java.util.function.BiPredicate;

import ghidra.app.script.GhidraScript;
import ghidra.features.bsim.gui.filters.*;
import ghidra.features.bsim.gui.search.results.BSimMatchResult;
import ghidra.features.bsim.gui.search.results.ExecutableResult;
import ghidra.features.bsim.query.FunctionDatabase;
import ghidra.features.bsim.query.FunctionDatabase.ErrorCategory;
import ghidra.features.bsim.query.description.FunctionDescription;
import ghidra.features.bsim.query.facade.*;
import ghidra.features.bsim.query.protocol.BSimFilter;
import ghidra.features.bsim.query.protocol.PreFilter;
import ghidra.program.database.symbol.FunctionSymbol;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.SourceType;
import ghidra.util.exception.CancelledException;

/**
 * Script showing how to apply filters to a BSim query. Currently we support three types
 * of filters, described below:
 * 
 * 	1. 	QUERY THRESHOLDS
 * 		These are the items at the top of the BSim query dialog:
 * 			Similarity
 * 			Confidence
 * 			Matches per Function
 * 		These are server-side filters that will be applied when the db is queried. 
 * 		
 * 2. 	PREFILTERS
 * 		Allows users to identify functions that meet certain criteria by specifying 
 * 		{@link BiPredicate}s. Any functions matching the predicate(s) will be included
 * 		in the result set.
 * 
 * 3. 	EXECUTABLE FILTERS
 * 		These are predefined filters that can be applied on the server or on the 
 * 		client (applied only to the results of a query). On the BSim query
 * 		dialog these are the items in the filter pulldown menu.
 * 		@see BSimFilterType
 *  
 *  SCRIPT FLOW
 *  	This example script does the following:
 *  
 *  	1) Set threshold filters
 *  	2) Set prefilters
 *  	3) Set executable filters
 *  	4) Query the database & print results
 *  	5) Set new executable filters
 *  	6) Print results
 *  
 *  NOTES:	1. 	You will be queried for the location of the BSim database. This URL
 *  			will take the form "ghidra://<ip address>/<database name>
 *  
 *  		2. 	This script is only an example - the specific filters demonstrated 
 *  			here will not necessarily apply to what's in your BSim database.
 *
 */
public class QueryWithFiltersScript extends GhidraScript {

	// Threshold settings.
	private static final int MAX_NUM_FUNCTIONS = 100;
	private static final double SIMILARITY_BOUND = 0.7;
	private static final double SIGNIFICANCE_BOUND = 0.0;

	// Restricts the number of results.
	private static final int NUM_EXES_TO_DISPLAY = 10;

	// Prefilter value we'll be setting.
	private static final double SELF_SIGNIFICANCE_BOUND = 40.0;

	private HashSet<FunctionSymbol> funcsToQuery;
	private SimilarFunctionQueryService queryService;
	private SFQueryInfo queryInfo;
	private BSimFilter bsimFilter;

	@Override
	protected void run() throws Exception {

		funcsToQuery = getFunctionsToQuery(currentProgram);
		queryService = new SimilarFunctionQueryService(currentProgram);
		queryInfo = new SFQueryInfo(funcsToQuery);
		bsimFilter = queryInfo.getBsimFilter();

		// Add threshold filters.
		queryInfo.setMaximumResults(MAX_NUM_FUNCTIONS);
		queryInfo.setSimilarityThreshold(SIMILARITY_BOUND);
		queryInfo.setSignificanceThreshold(SIGNIFICANCE_BOUND);

		// Add prefilters.
		setPrefilters();

		// Add a simple date filter.
		addBsimFilter(new DateLaterBSimFilterType(""), "01/01/1776");

		// Demonstration of a filter that allows for multiple entries. All filters but the 
		// DateEarlier and DateLater allow this. The effect is that each filter will be OR'd 
		// with the others. This is effectively the same as creating three distinct ArchEquals filters.
		//
		// ie: 	"The architecture can equal x86:LE:64:default OR the architecture can equal
		// 		ARM:LE_32:v4 OR ...."
		addBsimFilter(new ArchitectureBSimFilterType(),
			"x86:LE:64:default, x86:LE:32:default, ARM:LE:32:v4");

		// Another filter with multiple entries, but in this case since it is a "NotEqual" filter,
		// the items are "AND'd together.
		// 
		// ie: "The compiler cannot equal windows AND the compiler cannot equal foo_compiler".
		addBsimFilter(new NotCompilerBSimFilterType(), "windows, foo_compiler");

		//connect to the database
		try {
			String dbUrl =
				askString("", "Enter the URL of the BSim database:", "ghidra://localhost/bsimDb");
			queryService.initializeDatabase(dbUrl);
			FunctionDatabase.Error error = queryService.getLastError();
			if (error != null && error.category == ErrorCategory.Nodatabase) {
				println("Database [" + dbUrl + "] cannot be found (does it exist?)");
				return;
			}
		}
		catch (QueryDatabaseException e) {
			println(e.getMessage());
			return;
		}

		// Execute query and print results.
		List<BSimMatchResult> resultRows = executeQuery(queryInfo);
		printFunctionQueryResults(resultRows, "\nFunction-level results before filtering");

		// Add some simple post-query filters. These filters will only be applied to the result
		// set returned from the previous query.
		addBsimFilter(new Md5BSimFilterType(), currentProgram.getExecutableMD5());
		addBsimFilter(new CompilerBSimFilterType(), "gcc");
		//addBsimFilter(new FunctionTagBSimFilterType("KNOWN_LIBRARY", queryService), "false");

		// Apply the filters and print results.
		List<BSimMatchResult> filteredRows =
			BSimMatchResult.filterMatchRows(bsimFilter, resultRows);
		printFunctionQueryResults(filteredRows, "\nFunction-level results after filtering");
		printExecutableInformation(filteredRows);
	}

	@Override
	public void cleanup(boolean success) {
		if (queryService != null) {
			queryService.dispose();
		}
	}

	/***********************************************************************
	 * PRIVATE METHODS
	 ***********************************************************************/

	/**
	 * Adds a filter to the given filter container.
	 * 
	 * @param filterTemplate the filter type to add
	 * @param value the value of the filter
	 */
	private void addBsimFilter(BSimFilterType filterTemplate, String value) {
		String[] inputs = value.split(",");
		for (String input : inputs) {
			if (!input.trim().isEmpty()) {
				bsimFilter.addAtom(filterTemplate, input.trim());
			}
		}
	}

	/**
	 * Queries the database and returns the results. 
	 * 
	 * @param qInfo contains all information required for the query
	 * @return list of matches
	 * @throws QueryDatabaseException if there is a problem executing the query similar functions query
	 * @throws CancelledException if the user cancelled the operation
	 */
	private List<BSimMatchResult> executeQuery(SFQueryInfo qInfo)
			throws QueryDatabaseException, CancelledException {

		SFQueryResult queryResults = queryService.querySimilarFunctions(qInfo, null, monitor);
		List<BSimMatchResult> resultRows =
			BSimMatchResult.generate(queryResults.getSimilarityResults(), currentProgram);

		return resultRows;
	}

	/**
	 * Creates predicates that will be used to filter out functions. This example provides three
	 * different methods of doing this:
	 * 
	 * - anonymous class
	 * - lambda
	 * - static method
	 * 
	 * These are all possible because the filter takes a {@link BiPredicate}, which is a 
	 * functional interface.
	 * 
	 */
	private void setPrefilters() {

		PreFilter preFilter = queryInfo.getPreFilter();

		//
		// Option 1: 	Anonymous class
		//				Filters out any functions with a self significance less than a 
		//				certain value.
		//
		preFilter.addPredicate(new BiPredicate<Program, FunctionDescription>() {
			@Override
			public boolean test(Program t, FunctionDescription u) {
				return queryService.getLSHVectorFactory()
						.getSelfSignificance(
							u.getSignatureRecord().getLSHVector()) >= SELF_SIGNIFICANCE_BOUND;
			}
		});

		//
		// Option 2. 	Lambda expression
		//				Filters out any functions with a self significance less than a 
		//				certain value.
		//
		preFilter.addPredicate((x,
				y) -> queryService.getLSHVectorFactory()
						.getSelfSignificance(
							y.getSignatureRecord().getLSHVector()) >= SELF_SIGNIFICANCE_BOUND);

		//
		// Option 3. 	Static method
		//				Filters out any functions that are of type ANALYSIS.
		//
		preFilter.addPredicate(QueryWithFiltersScript::isNotAnalysisSourceType);
	}

	/**
	 * Returns a set of ALL functions (no stubs) in the given program.
	 * 
	 * @param program the program to get the functions from
	 * @return list of function symbols
	 */
	private HashSet<FunctionSymbol> getFunctionsToQuery(Program program) {
		HashSet<FunctionSymbol> functions = new HashSet<>();
		FunctionIterator fIter = program.getFunctionManager().getFunctionsNoStubs(true);
		for (Function func : fIter) {
			functions.add((FunctionSymbol) func.getSymbol());
		}
		return functions;
	}

	/**
	 * Returns true if the given function is NOT an analysis type.
	 * 
	 * @param program the current program
	 * @param funcDesc the function description object
	 * @return true if the symbol is NOT an analysis source type
	 */
	public static boolean isNotAnalysisSourceType(Program program, FunctionDescription funcDesc) {
		Address address =
			program.getAddressFactory().getDefaultAddressSpace().getAddress(funcDesc.getAddress());

		Function function = program.getFunctionManager().getFunctionAt(address);
		if (function == null || !function.getName().equals(funcDesc.getFunctionName())) {
			return false;
		}
		return function.getSymbol().getSource() != SourceType.ANALYSIS;
	}

	/**
	 * Prints a sorted list of executables represented in the function matches.
	 * 
	 * @param filteredRows list of function results
	 */
	private void printExecutableInformation(List<BSimMatchResult> filteredRows) {

		TreeSet<ExecutableResult> execrows = ExecutableResult.generateFromMatchRows(filteredRows);
		ExecutableResult[] results = new ExecutableResult[execrows.size()];
		results = execrows.toArray(results);

		Arrays.sort(results, new Comparator<ExecutableResult>() {
			@Override
			public int compare(ExecutableResult o1, ExecutableResult o2) {
				return Double.compare(o2.getSignificanceSum(), o1.getSignificanceSum());
			}
		});

		printf("Executable-level results:\n");
		for (int i = 0, max = Math.min(NUM_EXES_TO_DISPLAY, results.length); i < max; ++i) {
			printf("  MD5: %s\n", results[i].getExecutableRecord().getMd5());
			printf("  Executable Name: %s\n", results[i].getExecutableRecord().getNameExec());
			printf("  Function Count: %d\n", results[i].getFunctionCount());
			printf("  Significance Sum: %f\n\n", results[i].getSignificanceSum());
		}
	}

	/**
	 * Prints information about each function in the result set.
	 * 
	 * @param resultRows the list of rows containing the info to print
	 * @param title the title to print
	 */
	private void printFunctionQueryResults(List<BSimMatchResult> resultRows, String title) {
		printf(title + ": (%d)\n\n", resultRows.size());
		for (BSimMatchResult resultRow : resultRows) {
			printf("  queried function: %s\n",
				resultRow.getOriginalFunctionDescription().getFunctionName());
			printf("  matching function: %s\n",
				resultRow.getMatchFunctionDescription().getFunctionName());
			printf("  executable of matching function: %s\n",
				resultRow.getMatchFunctionDescription().getExecutableRecord().getNameExec());
			printf("  similarity: %f\n", resultRow.getSimilarity());
			printf("  significance: %f\n\n", resultRow.getSignificance());
		}
		printf("\n");
	}

}
